text/E10_ABM/ABM.tex

```diff
@@ -0,0 +1,88 @@
+
+\begin{algorithm}[H]
+    \caption{$k$ - przybliżone wyszukiwanie wzorca w tekście względem odległości edycyjnej}
+    \Input{Słowa $t, p \in \A^+$ ($|t| = n$, $|p| = m$)}
+    \Output{Zbiór liczb $S = \{j : \text{ istnieje podsłowo T zakończone na $t_j$ o odległości edycyjnej od p równej co najwyżej $k$}\}$.}
+\end{algorithm}
+
+\subsection{Algorytm programowania dynamicznego.}
+Podstawowym algorytmem rozwiązującym zadany wyżej problem jest następujący algorytm programowania dynamicznego wyliczający tablicę $D(i,j)$ oznaczającą minimalną odległość edycyjną między słowami $p_1...p_i$ oraz dowolnym podsłowem $T$ zakończonym na $t_j$:
+\begin{align*}
+    & D(0,j) = 0,\ 0 \leq j \leq n \\
+    & D(i, j) = min\ \left\{\begin{array}{ll}
+         &  D(i-1,j)+1\\
+         &  D(i-1,j-1) + \text{ if } p_i = t_j \text{ then } 0 \text{ else } 1 \\
+         &  D(i, j-1)+1
+    \end{array} \right.
+\end{align*}
+Rozwiązaniem są wszystkie $j$ takie, że $D(m,j) \leq k$. Algorytm ten działa w czasie $O(mn)$.
+
+\subsection{Idea algorytmu ABM (Approximate Boyer-Moore)}
+
+Algorytm oparty na pomyśle analogicznym do algorytmu Boyer'a-Moore'a będzie będzie działać w dwóch głównych fazach: \textit{skanowania} i \textit{sprawdzania}. 
+W fazie skanowania przechodzimy przez dane słowo $t$ i zaznaczamy pewne komórki pierwszego rzędu tablicy $D$ (tj komórki postaci $D(0,j)$). 
+Po zakończeniu fazy skanowania rozpoczynamy fazę sprawdzania, czyli dla każdej zaznaczonej komórki $D$ będziemy obliczali przekątną tablicy $D$ poprzez zwykłe programowanie dynamiczne. 
+Kiedykolwiek nasza procedura dynamiczna będzie odnosić się do niewyliczonej komórki możemy przyjąć, że jej wartość wynosi $\infty.$\\
+Faza sprawdzania wydaje się dość intuicyjna dlatego skupmy się na fazie skanowania.\\
+
+Faza skanowania powtarza w pętli dwie czynności. \textit{Zaznacza} (mark) i \textit{przesuwa} (shift) indeks który skanujemy. Operacja przesuwania jest analogiczna do operacji przesuwania w algorytmie $Boyer'a-Moore'a$. Operacja zaznaczania wykonywana jest wtedy gdy obecny indeks wymaga dokładniejszego sprawdzenia przez wyliczenie tablicy $D.$
+
+\subsection{Faza skanowania algorytmu ABM}
+\begin{definition}{}{}
+    Mówimy, że dla każdego $D(i,j)$ istnieje {\bf \textit{łuk minimalizujący}} z $D(i-1, j)$ do $D(i, j)$ jeśli $D(i,j) = D(i - 1, j)+1$, z $D(i, j-1)$ do $D(i, j)$ jeśli $D(i,j) = D(i, j-1)+1$, oraz z $D(i-1,j-1)$ do $D(i, j)$ jeśli $D(i,j) = D(i-1, j-1)$ gdy $p_i = t_j$ albo 
+    $D(i,j) = D(i-1, j-1) + 1$ gdy $p_i \neq t_j.$
+\end{definition}
+\begin{definition}{}{}
+    {\bf \textit{Minimalizującą ścieżką}} nazywamy dowolną ścieżkę złożoną z minimalizujących łuków zaczynającą w $D(0,j)$ w pierwszym rzędzie tablicy $D$ do komórki $D(m, h)$ znajdującej się w ostatnim rzędzie. Minimalizującą ścieżkę nazywamy {\bf \textit{dobrą}} gdy prowadzi do $D(m, h) \leq k.$
```

Comment:
Please add " for all $1 \le h \le n$". I know it's straightfoward, but currently it may be ambigious

Reply:
Done.